core/src/main/kotlin/com/strumenta/kolasu/symbolresolution/Testing.kt

```diff
@@ -5,24 +5,57 @@ import com.strumenta.kolasu.model.PossiblyNamed
 import com.strumenta.kolasu.model.ReferenceByName
 import com.strumenta.kolasu.model.nodeProperties
 import com.strumenta.kolasu.traversing.walk
+import kotlin.reflect.KType
 import kotlin.reflect.KTypeProjection
 import kotlin.reflect.KVariance
 import kotlin.reflect.full.createType
 import kotlin.reflect.full.isSubtypeOf
 import kotlin.test.assertTrue
 
 fun Node.assertAllReferencesResolved() {
-    this.walk().forEach {
+    this.assertAllReferencesOfTypeResolved<PossiblyNamed>()
+}
+
+fun Node.assertNotAllReferencesResolved() {
+    this.assertNotAllReferencesOfTypeResolved<PossiblyNamed>()
+}
+
+inline fun <reified T : PossiblyNamed> Node.assertAllReferencesOfTypeResolved() {
+    assertTrue { this.getReferenceResolutionInfo<T>().all { it } }
+}
+
+inline fun <reified T : PossiblyNamed> Node.assertNotAllReferencesOfTypeResolved() {
+    assertTrue { this.getReferenceResolutionInfo<T>().ifEmpty { sequenceOf(false) }.any { !it } }
+}
+
+fun Node.assertAllReferencesOfPropertyResolved(targetProperty: ReferenceByNameProperty) {
+    assertTrue { this.getReferenceResolutionInfo(targetProperty).all { it } }
+}
+
+fun Node.assertNotAllReferencesOfPropertyResolved(targetProperty: ReferenceByNameProperty) {
+    assertTrue { this.getReferenceResolutionInfo(targetProperty).ifEmpty { sequenceOf(false) }.any { !it } }
+}
+
+inline fun <reified T : PossiblyNamed> Node.getReferenceResolutionInfo(): Sequence<Boolean> {
+    return this.walk().flatMap {
         it.nodeProperties
-            .filter { property ->
-                property.returnType.isSubtypeOf(
-                    ReferenceByName::class.createType(
-                        arguments = listOf(
-                            KTypeProjection(variance = KVariance.OUT, type = PossiblyNamed::class.createType()),
-                        ),
-                    ),
-                )
-            }
-            .forEach { property -> assertTrue { (property.get(it) as ReferenceByName<*>).resolved } }
+            .filter { property -> property.returnType.isSubtypeOf(referenceByNameType<T>()) }
+            .mapNotNull { property -> property.get(it) }
+            .map { value -> (value as ReferenceByName<*>).resolved }
     }
 }
+
+fun Node.getReferenceResolutionInfo(targetProperty: ReferenceByNameProperty): Sequence<Boolean> {
```

Comment:
I do not understand the goal of this operation by looking at its name

Reply:
This is used to retrieve all values of the `resolved` field for each `ReferenceByName` of the given `targetProperty`. Actually, it is only used internally and not meant to be used outside. Perhaps, it would be better to make it private, what do you think? The other variations of this function might need to be inlined as currently can not be defined as private - the compiler would give error as private inline functions can not be used in public functions.